blocks/utils/containers.py

```diff
@@ -0,0 +1,57 @@
+from abc import ABCMeta
+from collections import MutableSequence
+
+from six import add_metaclass
+
+
+@add_metaclass(ABCMeta)
+class AnnotatingList(MutableSequence):
+    """Mutable sequence performing operations on inserted/removed items.
+
+    Parameters
+    ----------
+    items : iterable, optional
+        An iterable of items to initialize the sequence with.
+
+    """
+    def __init__(self, items=None):
+        self._items = []
+        if not items:
+            items = []
+        for item in items:
+            self.append(item)
+
+    def __repr__(self):
+        return repr(self._items)
+
+    def __eq__(self, other):
+        return self._items == other
+
+    def __ne__(self, other):
+        return self._items != other
+
+    def __getitem__(self, key):
+        return self._items[key]
+
+    def _setitem(self, key, value):
+        """The operation to perform when an item is inserted/appended."""
+        pass
+
+    def _delitem(self, key):
```

Comment:
Actually I can see this one perhaps not being annotated as `@abstractmethod`, but `_setitem` seems fundamental.